lib/sendstate.c

```diff
@@ -125,7 +125,7 @@ int quicly_sendstate_acked(quicly_sendstate_t *state, quicly_sendstate_sent_t *a
         if ((ret = quicly_ranges_subtract(&state->pending, args->start, args->end)) != 0)
             return ret;
     }
-    assert(state->pending.num_ranges == 0 || state->acked.ranges[0].end <= state->pending.ranges[0].start);
+    assert(state->final_size == UINT64_MAX || state->pending.num_ranges == 0 || state->acked.ranges[0].end <= state->pending.ranges[0].start);
```

Comment:
Could you please elaborate how this assert could be a problem? EOS is intended to be represented as a virtual byte at offset `final_size`, see [the use of `final_size + 1` in quicly_sendstate_shutdown](https://github.com/h2o/quicly/blob/master/lib/sendstate.c#L74-L79).

Reply:
Err, yes, sorry about that. I've amended the check. My intention was to only do this check if we haven't closed, since otherwise, it seems possible to have pending.num_ranges == 1 and state->acked.ranges[0].end > state->pending.ranges[0].start

Reply:
I'm not sure if I follow.

As the CI failure indicates, this code path is visited regardless of `final_size` being UINT64_MAX or not.

To restate my point, "ranges" represent the byte ranges being set, where FIN is represented as a virtual character at the end. Therefore, this assertion does not need to depend on the value of `final_size`. Assertion failure here indicates that there is likely be a bug elsewhere.

Reply:
I see, thank you! I'll keep digging.